File: 59-spiral-matrix-ii/59-spiral-matrix-ii.java
class Solution {
    public int[][] generateMatrix(int n) {
        int spiral[][] = new int[n][n];
        int cnt = 1;
        for(int layer = 0;layer < (n+1)/2; layer++){
            
            for(int i=layer;i<(n - layer);i++){
                spiral[layer][i] = cnt++;
            }
            
            for(int i=layer + 1;i<(n - layer);i++){
                spiral[i][n-layer-1] = cnt++;
            }
            
            for(int i=n-layer-2;i>=layer;i--){
                spiral[n-layer-1][i] = cnt++;
            }
            
            for(int i=n-layer-2;i>layer;i--){
                spiral[i][layer] = cnt++;
            }
        }
        return spiral;
    }
}

File: 59-spiral-matrix-ii/59-spiral-matrix-ii.cpp
class Solution {
public:
    int floorMod(int x,int y) {
        return ((x%y) + y) % y;
    }
    vector<vector<int>> generateMatrix(int n) {
        vector<vector<int>> v(n,vector<int>(n,0));
        
        int dir[4][2] = {{0,1},{1,0},{0,-1},{-1,0}};
        
        int d = 0,cnt = 1;
        int row = 0,col = 0;
        

        while(cnt <= n * n){
            v[row][col] = cnt++;
            int r = floorMod(row + dir[d][0],n);
            int c = floorMod(col + dir[d][1],n);
            
            if(v[r][c] != 0) d = (d+1)%4;
            
            row += dir[d][0];
            col += dir[d][1];
        }
        return v;
    }
};